### PR TITLE
feat: fix the logic short-circuit bug

### PR DESCRIPTION
**Pull request by tx2002**
For: https://github.com/casbin/node-casbin/issues/493

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece @Shivansh-yadav13 please review

File: index.ts [hunk 1]
```diff
@@ -150,10 +150,11 @@ function evaluate(_node: jsep.Expression, context: object) {
...
-        return evaluate(node.left, context) || evaluate(node.right, context);
-      } else if (node.operator === '&&') {
-        return evaluate(node.left, context) && evaluate(node.right, context);
+      const leftValue = evaluate(node.left, context);
+      if (node.operator === '||' && leftValue) {
+        return leftValue;
+      } else if (node.operator === '&&' && !leftValue) {
+        return leftValue;
       }
       return binops[node.operator](evaluate(node.left, context), evaluate(node.right, context));
```
Comment (nodece) [inline index.ts]:
```suggestion
      return binops[node.operator](leftValue, evaluate(node.right, context));
```

File: index.ts [hunk 2]
```diff
@@ -222,15 +223,19 @@ async function evalAsync(_node: jsep.Expression, context: object) {
...
-        return (
-          (await evalAsync(node.left, context)) ||
-          (await evalAsync(node.right, context))
-        );
+        const left = await evalAsync(node.left, context);
+        if (left) {
+          return left;
+        }
+        const right = await evalAsync(node.right, context);
+        return right;
```
Comment (nodece) [inline index.ts]:
```suggestion
        return await evalAsync(node.right, context);
```

File: index.ts [hunk 3]
```diff
@@ -222,15 +223,19 @@ async function evalAsync(_node: jsep.Expression, context: object) {
...
-        return (
-          (await evalAsync(node.left, context)) &&
-          (await evalAsync(node.right, context))
-        );
+        const left = await evalAsync(node.left, context);
+        if (!left) {
+          return left;
+        }
+        const right = await evalAsync(node.right, context);
+        return right;
```
Comment (nodece) [inline index.ts]:
```suggestion
        return await evalAsync(node.right, context);
```

File: test.js [hunk 4]
```diff
@@ -62,9 +62,9 @@ const fixtures = [
   {expr: '1 >= 2',          expected: false },
 
   // logical expression lazy evaluation
-  {expr: 'true || false',  expected: true  },
+  {expr: 'true || throw()',  expected: true  },
```
Comment (nodece) [inline test.js:65]:
```suggestion
  {expr: 'true || false',  expected: true  },
  {expr: 'true || throw()',  expected: true  },
```

File: test.js [hunk 5]
```diff
@@ -62,9 +62,9 @@ const fixtures = [
...
   {expr: 'false || true',    expected: true  },
-  {expr: 'false && true', expected: false  },
+  {expr: 'false && throw()', expected: false  },
```
Comment (nodece) [inline test.js:67]:
```suggestion
  {expr: 'false && true', expected: false  },
  {expr: 'false && throw()', expected: false  },
```

**Review by nodece — CHANGES_REQUESTED** [review]

File: test.js (hunk 4 above)
Comment (tx2002) [reply test.js:65]: This throw() is used to test the short-circuit logic and will not be called.
see the same in [donmccurdy/expression-eval](https://github.com/donmccurdy/expression-eval/blob/3ee41060d357887d01beb65a6f3ff2c5a27143f6/test.js#L66)

File: test.js (hunk 5 above)
Comment (tx2002) [reply test.js:67]: the same

File: test.js (hunk 4 above)
Comment (nodece) [reply test.js:65]: I see, but you still need to keep the original test cases.

Comment (tx2002) [reply test.js:65]: Could you please tell me why

Comment (nodece) [reply test.js:65]: Just to cover all cases.

**Pull request merged**